{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3d9175b3",
   "metadata": {},
   "source": [
    "### Rosalind RNA Splicing\n",
    "Given: A DNA string s (of length at most 1 kbp) and a collection of substrings of s acting as introns. All strings are given in FASTA format.\n",
    "\n",
    "Return: A protein string resulting from transcribing and translating the exons of s (Note: Only one solution will exist for the dataset provided.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9b9022d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "codons = { \"UUU\": \"F\",\n",
    "      \"CUU\": \"L\",     \n",
    "      \"AUU\": \"I\",      \n",
    "      \"GUU\": \"V\",\n",
    "      \"UUC\": \"F\" ,     \n",
    "      \"CUC\": \"L\", \n",
    "      \"AUC\": \"I\",      \n",
    "      \"GUC\": \"V\",\n",
    "      \"UUA\": \"L\",   \n",
    "      \"CUA\": \"L\",      \n",
    "      \"AUA\": \"I\",\n",
    "      \"GUA\": \"V\",\n",
    "      \"UUG\": \"L\",   \n",
    "      \"CUG\": \"L\",\n",
    "      \"AUG\": \"M\", \n",
    "      \"GUG\": \"V\",\n",
    "      \"UCU\": \"S\",   \n",
    "      \"CCU\": \"P\",     \n",
    "      \"ACU\": \"T\",\n",
    "      \"GCU\": \"A\",\n",
    "      \"UCC\": \"S\",\n",
    "      \"CCC\": \"P\",\n",
    "      \"ACC\": \"T\",\n",
    "      \"GCC\": \"A\",\n",
    "      \"UCA\": \"S\",\n",
    "      \"CCA\": \"P\",\n",
    "      \"ACA\": \"T\",    \n",
    "      \"GCA\": \"A\",\n",
    "      \"UCG\": \"S\",\n",
    "      \"CCG\": \"P\", \n",
    "      \"ACG\": \"T\",\n",
    "      \"GCG\": \"A\",\n",
    "      \"UAU\": \"Y\",\n",
    "      \"CAU\": \"H\",\n",
    "      \"AAU\": \"N\",\n",
    "      \"GAU\": \"D\",\n",
    "      \"UAC\": \"Y\",\n",
    "      \"CAC\": \"H\",\n",
    "      \"AAC\": \"N\",\n",
    "      \"GAC\": \"D\",\n",
    "      \"UAA\": \" \", #stop codon\n",
    "      \"CAA\": \"Q\", \n",
    "      \"AAA\": \"K\", \n",
    "      \"GAA\": \"E\",\n",
    "      \"UAG\": \" \", #stop codon\n",
    "      \"CAG\": \"Q\",\n",
    "      \"AAG\": \"K\", \n",
    "      \"GAG\": \"E\",\n",
    "      \"UGU\": \"C\",\n",
    "      \"CGU\": \"R\",\n",
    "      \"AGU\": \"S\",\n",
    "      \"GGU\": \"G\",\n",
    "      \"UGC\": \"C\",\n",
    "      \"CGC\": \"R\",\n",
    "      \"AGC\": \"S\",\n",
    "      \"GGC\": \"G\",\n",
    "      \"UGA\": \" \", #stop codon\n",
    "      \"CGA\": \"R\",\n",
    "      \"AGA\": \"R\",\n",
    "      \"GGA\": \"G\",\n",
    "      \"UGG\": \"W\", \n",
    "      \"CGG\": \"R\", \n",
    "      \"AGG\": \"R\",\n",
    "      \"GGG\": \"G\"}\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3328740a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(file_path):\n",
    "    DNA = {}\n",
    "    current_id = None\n",
    "    current_seq = [] \n",
    "    with open(file_path, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if current_id is not None:\n",
    "                    DNA[current_id] = \"\".join(current_seq)\n",
    "                current_id = line[1:]\n",
    "                current_seq = []\n",
    "            else:\n",
    "                current_seq.append(line)\n",
    "        if current_id is not None:\n",
    "            DNA[current_id] = \"\".join(current_seq)\n",
    "    return DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c833ef8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Rosalind = \"ATGGTCTACATAGCTGACAAACAGCACGTAGCAATCGGTCGAATCTCGAGAGGCATATGGTCACATGATCGGTCGAGCGTGTTTCAAAGTTTGCGCCTAG\"\n",
    "introns = ['ATCGGTCGAA', 'ATCGGTCGAGCGTGT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7cec50a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGGTCTACATAGCTGACAAACAGCACGTAGCATCTCGAGAGGCATATGGTCACATGTTCAAAGTTTGCGCCTAG\n",
      "100\n",
      "75\n",
      "AUGGUCUACAUAGCUGACAAACAGCACGUAGCAUCUCGAGAGGCAUAUGGUCACAUGUUCAAAGUUUGCGCCUAG\n"
     ]
    }
   ],
   "source": [
    "exons = Rosalind\n",
    "for intron in introns:\n",
    "    exons = exons.replace(intron,\"\")\n",
    "print(exons)\n",
    "print(len(Rosalind))\n",
    "print(len(exons))\n",
    "\n",
    "mrna = exons.replace('T', 'U')\n",
    "print(mrna)\n",
    "\n",
    "start_index = mrna.find(\"AUG\")\n",
    "if start_index == -1:\n",
    "    print('no start codon found')  \n",
    "\n",
    "mrna = mrna[start_index:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1a30cfc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUG GUC UAC AUA GCU GAC AAA CAG CAC GUA GCA UCU CGA GAG GCA UAU GGU CAC AUG UUC AAA GUU UGC GCC UAG\n"
     ]
    }
   ],
   "source": [
    "triple = ''\n",
    "for i in range(0, len(mrna), 3):   \n",
    "    codon = mrna[i:i+3]\n",
    "    if len(codon) < 3:   \n",
    "        break\n",
    "    if codon in ('UAA', 'UAG', 'UGA'):  \n",
    "        triple = triple + ' ' + codon\n",
    "        break\n",
    "    if triple == '':\n",
    "        triple = codon\n",
    "    else:\n",
    "        triple = triple + \" \" + codon\n",
    "print(triple)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "69b54ab3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MVYIADKQHVASREAYGHMFKVCA \n"
     ]
    }
   ],
   "source": [
    "protein = \"\"\n",
    "for key in triple.split(\" \"):\n",
    "    protein = protein + codons[key]\n",
    "print(protein)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9ccd437",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MVYIADKQHVASREAYGHMFKVCA '"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f6f6c52",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e8bb205",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b81a3d26",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "carpentries",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
